### PR TITLE
add d.ts generation with typescript compiler

### PR DESCRIPTION
**Pull request by qballer**
This reverts commit 82fe443651272b3881ac43d52532bc24f5c8087c.

## Proposed Changes

- Type generation (d.ts)
- Some type fixes

---
This change is [Reviewable](https://reviewable.io/reviews/teambit/bit/2123)

### CODE REVIEW

**Comment by qballer** [comment]
@GiladShoham the e2e also seem to be failing on master. I wonder why

File: src/scope/objects/repository.ts [hunk 1]
```diff
@@ -115,9 +115,8 @@ export default class Repository {
     const refs = await this.listRefs();
     return Promise.all(refs.map(ref => this.load(ref)));
   }
-
-  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
-  async listRefs(): Promise<Refs[]> {
+  // type any
+  async listRefs(): Promise<Array<any>> {
```
Comment (davidfirst) [inline src/scope/objects/repository.ts]: @qballer , it needs to be `Ref[]`, it was probably a typo before (`Refs[]` with `s`).

File: src/scope/repositories/sources.ts [hunk 2]
```diff
@@ -182,8 +182,7 @@ to quickly fix the issue, please delete the object at "${this.objects().objectPa
     flattenedTesterDependencies,
     specsResults
   }: {
-    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
-    consumerComponent: $ReadOnly<ConsumerComponent>;
+    consumerComponent: ConsumerComponent;
```
Comment (davidfirst) [inline src/scope/repositories/sources.ts]: @qballer , why did you remove the ReadOnly? it's deliberately there.

Comment (qballer) [reply src/scope/repositories/sources.ts]: It isn't respected in the function, and ReadOnly breaks. $ReadOnly is no good here.

File: src/scope/objects/repository.ts (hunk 1 above)
Comment (qballer) [reply src/scope/objects/repository.ts]: Cool, will change that.

File: src/scope/repositories/sources.ts (hunk 2 above)
Comment (davidfirst) [reply src/scope/repositories/sources.ts]: @qballer , The idea behind it was to not change anything in the parameter instance. All changes must be done on a clone of the parameter.
So, the syntax you see here is from Flow. The Typescript syntax is different, but it's still possible.
You can change the entire line to:
`readonly consumerComponent: ConsumerComponent;`
In general, you can see this great reference of the flow vs typescript in terms of syntax changes: https://github.com/niieani/typescript-vs-flowtype

Comment (qballer) [reply src/scope/repositories/sources.ts]: I swear something violated it. made it readonly.

File: src/consumer/component/dependencies/dependency-resolver/dependencies-resolver.ts [hunk 3]
```diff
@@ -55,7 +55,7 @@ export default class DependencyResolver {
   allDependencies: AllDependencies;
   allPackagesDependencies: AllPackagesDependencies;
   // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
-  issues: $PropertyType<Component, 'issues'>;
+  issues: any; //$PropertyType<Component, 'issues'>;
```
Comment (davidfirst) [inline src/consumer/component/dependencies/dependency-resolver/dependencies-resolver.ts]: @qballer  , in Typescript it's actually easier, it should be `Component['issues']`

**Comment by GiladShoham** [comment]
@davidfirst it's already merged.. feel free to change it:)
